pkg/buildkit/worker/pool.go

```diff
@@ -79,72 +84,85 @@ func NewPool(ctx context.Context, clientset kubernetes.Interface, conf config.Bu
 		o = fn(o)
 	}
 
-	scaler := NewStatefulSetScaler(ScalerOpt{
-		Name:                conf.StatefulSetName,
-		Namespace:           conf.Namespace,
-		Clientset:           clientset,
-		WatchTimeoutSeconds: o.watchTimeoutSeconds,
-		Log:                 o.log.WithName("statefulset-scaler"),
-	})
-
-	fs := fields.SelectorFromSet(map[string]string{"status.phase": "Running"})
 	ls := labels.SelectorFromSet(conf.PodLabels)
-	podListOptions := metav1.ListOptions{
-		LabelSelector: ls.String(),
-		FieldSelector: fs.String(),
-	}
+	podListOptions := metav1.ListOptions{LabelSelector: ls.String()}
 
 	ctx, cancel := context.WithCancel(ctx)
 	wp := &workerPool{
-		ctx:             ctx,
-		cancel:          cancel,
-		log:             o.log,
-		poolSyncTime:    o.syncWaitTime,
-		podMaxIdleTime:  o.maxIdleTime,
-		uuid:            string(uuid.NewUUID()),
-		requests:        NewRequestQueue(),
-		results:         make(chan PodRequestResult),
-		podClient:       clientset.CoreV1().Pods(conf.Namespace),
-		endpointsClient: clientset.CoreV1().Endpoints(conf.Namespace),
-		podListOptions:  podListOptions,
-		serviceName:     conf.ServiceName,
-		servicePort:     conf.DaemonPort,
-		namespace:       conf.Namespace,
-		scaler:          scaler,
+		ctx:               ctx,
+		cancel:            cancel,
+		log:               o.log,
+		poolSyncTime:      o.syncWaitTime,
+		podMaxIdleTime:    o.maxIdleTime,
+		uuid:              string(newUUID()),
+		requests:          NewRequestQueue(),
+		notifyReconcile:   make(chan struct{}, 1),
+		podClient:         clientset.CoreV1().Pods(conf.Namespace),
+		endpointsClient:   clientset.CoreV1().Endpoints(conf.Namespace),
+		podListOptions:    podListOptions,
+		serviceName:       conf.ServiceName,
+		servicePort:       conf.DaemonPort,
+		statefulSetName:   conf.StatefulSetName,
+		statefulSetClient: clientset.AppsV1().StatefulSets(conf.Namespace),
+		namespace:         conf.Namespace,
 	}
 
-	go wp.processRequests()
-	go wp.monitorPool()
+	wp.log.Info("Starting worker pod monitor", "syncTime", wp.poolSyncTime.String())
+	go func() {
+		ticker := time.NewTicker(wp.poolSyncTime)
+		defer ticker.Stop()
+		defer close(wp.notifyReconcile)
+
+		for {
+			if err := wp.updateWorkers(wp.ctx); err != nil {
+				wp.log.Error(err, "Failed to update worker pool")
+			}
+
+			select {
+			case <-wp.notifyReconcile:
+			case <-ticker.C:
+				// break out of the select, let it update on the loop
+			case <-wp.ctx.Done():
+				wp.log.Info("Shutting down worker pod monitor")
+				for wp.requests.Len() > 0 {
+					close(wp.requests.Dequeue())
+				}
+
+				return
+			}
+		}
+	}()
 
 	return wp
 }
 
 func (p *workerPool) Get(ctx context.Context) (string, error) {
 	ch := make(chan PodRequestResult, 1)
-	defer close(ch)
 
 	p.requests.Enqueue(ch)
+	defer p.requests.Remove(ch)
```

Comment:
when is this channel closed? the line below

```go
		// when the channel is closed, we receive nil
		if result.pod != nil {
			return p.buildEndpointURL(ctx, result.pod.Name)
		}
```

indicates we need to guard against a close channel, but i don't see anywhere where we close these channels.